test/unit/connection/error-tests.js

```diff
@@ -37,26 +37,52 @@ suite.test('connection does not emit ECONNRESET errors during disconnect', funct
   done()
 })
 
+var SSLNegotiationPacketTests = [
+  {
+    testName: 'connection does not emit ECONNRESET errors during disconnect also when using SSL',
+    errorMessage: null,
+    response: 'S',
+    responseType: 'sslconnect'
+  },
+  {
+    testName: 'connection emits an error when SSL is not supported',
+    errorMessage: 'The server does not support SSL connections',
+    response: 'N',
+    responseType: 'error'
+  },
+  {
+    testName: 'connection emits an error when postmaster responds to SSL negotiation packet',
+    errorMessage: 'There was an error establishing an SSL connection',
+    response: 'E',
+    responseType: 'error'
+  }
+]
 
-suite.test('connection does not emit ECONNRESET errors during disconnect also when using SSL', function (done) {
-  // our fake postgres server, which just responds with 'S' to start SSL
-  var socket
-  var server = net.createServer(function (c) {
-    socket = c
-    c.once('data', function (data) {
-      c.write(Buffer.from('S'))
+for (var i = 0; i < SSLNegotiationPacketTests.length; i++) {
+  var tc = SSLNegotiationPacketTests[i]
+  suite.test(tc.testName, function (done) {
+    // our fake postgres server
+    var socket
+    var server = net.createServer(function (c) {
+      socket = c
+      c.once('data', function (data) {
+        c.write(Buffer.from(tc.response))
+      })
     })
-  })
 
-  server.listen(7778, function () {
-    var con = new Connection({ssl: true})
-    con.connect(7778, 'localhost')
-    assert.emits(con, 'sslconnect', function () {
-      con.end()
-      socket.destroy()
-      server.close()
-      done()
+    server.listen(7778, function () {
+      var con = new Connection({ssl: true})
+      con.connect(7778, 'localhost')
+      assert.emits(con, tc.responseType, function (err) {
+        if (err) {
```

Comment:
Should this run `if (tc.errorMessage !== null)`?

Reply:
If there is no `errorMessage`, then no. Good catch.

Reply:
Ah, I meant that this won’t check for an expected error if an error isn’t emitted.

Reply:
Ah, yeah, I'm with you. That's some short sightedness on my part – I was checking for the other case (when an error is emitted and one isn't expected). I've now modified this test to run if either an `errorMessage` is specified (for expected errors), or if `err` is present (for unexpected errors).